### PR TITLE
Support for NGSIv2

### PR DESCRIPTION
**Pull request by dcalvoalonso**
Fixes: [527](https://github.com/telefonicaid/iotagent-node-lib/issues/527)

First version that supports NGSIv2. 
Some aspects are not covered yet:
* Registrations to CB : specified in the last version of NGSI but not implemented in CB)
* Acts as Context Provider for queries/updates coming from CB (for passive attributes and commands): Not specified yet

### CODE REVIEW

File: lib/fiware-iotagent-lib.js [hunk 1]
```diff
@@ -177,20 +199,115 @@ function deactivate(callback) {
...
+exports.unregister = unregister;
+exports.query = query;
+exports.update = update;
+exports.setCommandResult = setCommandResult;
+exports.listDevices = listDevices;
+exports.getDevice = getDevice;
+exports.getDeviceByName = getDeviceByName;
+exports.getDevicesByAttribute = getDevicesByAttribute;
+exports.mergeDeviceWithConfiguration = mergeDeviceWithConfiguration;
+exports.retrieveDevice = retrieveDevice;
```
Comment (jmcanterafonseca) [inline lib/fiware-iotagent-lib.js]: are these changes really needed?

File: lib/fiware-iotagent-lib.js [hunk 2]
```diff
@@ -177,20 +199,115 @@ function deactivate(callback) {
     contextServer.stop(callback);
 }
 
+function update(entityName, attributes, typeInformation, token, callback) {
+    ngsi.update(entityName, attributes, typeInformation, token, callback);
+}
+
+function query(entityName, attributes, typeInformation, token, callback) {
+    ngsi.query(entityName, attributes, typeInformation, token, callback);
+
```
Comment (jmcanterafonseca) [inline lib/fiware-iotagent-lib.js]: not needed line, please use a Linter

File: lib/fiware-iotagent-lib.js [hunk 3]
```diff
@@ -177,20 +199,115 @@ function deactivate(callback) {
...
+function setCommandResult(entityName, resource, apikey, commandName,
+                          commandResult, status, deviceInformation, callback) {
+    ngsi.setCommandResult(entityName, resource, apikey, commandName,
+                          commandResult, status, deviceInformation, callback);
+
+}
+
+function addUpdateMiddleware(middleware) {
+    ngsi.addUpdateMiddleware(middleware);
+
```
Comment (jmcanterafonseca) [inline lib/fiware-iotagent-lib.js]: not needed line

File: lib/fiware-iotagent-lib.js [hunk 4]
```diff
@@ -199,23 +316,23 @@ exports.setDataQueryHandler = contextServer.setQueryHandler;
 exports.setConfigurationHandler = contextServer.setConfigurationHandler;
 exports.setProvisioningHandler = contextServer.setProvisioningHandler;
 exports.setNotificationHandler = contextServer.setNotificationHandler;
-exports.addUpdateMiddleware = ngsi.addUpdateMiddleware;
-exports.addQueryMiddleware = ngsi.addQueryMiddleware;
+exports.addUpdateMiddleware = addUpdateMiddleware;
```
Comment (jmcanterafonseca) [inline lib/fiware-iotagent-lib.js]: are these changes really needed?

File: lib/fiware-iotagent-lib.js [hunk 5]
```diff
@@ -235,6 +352,15 @@ exports.dataPlugins = {
     bidirectionalData: require('./plugins/bidirectionalData')
 };
 
+exports.dataPluginsNgsi2 = {
+    attributeAlias: require('./pluginsNgsi2/attributeAlias'),
```
Comment (jmcanterafonseca) [inline lib/fiware-iotagent-lib.js]: for NGSIv2 you don't need to create a new plugin

File: lib/pluginsNgsi2/addEvent.js [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+ * See the GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public
+ * License along with fiware-iotagent-lib.
+ * If not, seehttp://www.gnu.org/licenses/.
+ *
+ * For those usages not covered by the GNU Affero General Public License
+ * please contact with::daniel.moranjimenez@telefonica.com
+ *
+ * Modified work Copyright 2017 Atos Spain S.A
```
Comment (jmcanterafonseca) [inline lib/pluginsNgsi2/addEvent.js]: this is not the way to state your authorship. You can add a sentence like

`Author: Daniel Calvo Atos Spain S.A. ` 

IPR issues are currently being discussed by Fiware Foundation and current IPR owners

File: lib/pluginsNgsi2/attributeAlias.js [hunk 7]
```diff
@@ -0,0 +1,111 @@
+/*
```
Comment (jmcanterafonseca) [inline lib/pluginsNgsi2/attributeAlias.js]: do we really need this new component?

File: lib/pluginsNgsi2/bidirectionalData.js [hunk 8]
```diff
@@ -0,0 +1,276 @@
+/*
```
Comment (jmcanterafonseca) [inline lib/pluginsNgsi2/bidirectionalData.js]: do we really need this new component?

File: lib/fiware-iotagent-lib.js (hunk 2 above)
Comment (mrutid) [reply lib/fiware-iotagent-lib.js]: Indeed, not needed code... why this? (and all the rest, seems pretty much out of scope of the PR)

File: lib/pluginsNgsi2/addEvent.js (hunk 6 above)
Comment (fgalan) [reply lib/pluginsNgsi2/addEvent.js]: An example (currently used at Orion repository) of the style proposed by @jmcanterafonseca:

https://github.com/telefonicaid/fiware-orion/blob/master/src/lib/serviceRoutinesV2/optionsGetOnly.cpp#L23

File: lib/pluginsNgsi2/expressionParser.js [hunk 9]
```diff
@@ -0,0 +1,202 @@
+/*
```
Comment (jmcanterafonseca) [inline lib/pluginsNgsi2/expressionParser.js]: is this new code? or just moving an existing component?

File: lib/pluginsNgsi2/expressionPlugin.js [hunk 10]
```diff
@@ -0,0 +1,86 @@
+
```
Comment (jmcanterafonseca) [inline lib/pluginsNgsi2/expressionPlugin.js]: is the expression plugin different than the NGSIv1 one?

File: lib/pluginsNgsi2/expressionPlugin.js [hunk 11]
```diff
@@ -0,0 +1,86 @@
...
+
+function update(entity, typeInformation, callback) {
+    try {
+        var entityAtts = [];
+        var expressionAttributes = [];
+        entityAtts = utils.createArrayFromAttributes(entity);
+        var context = parser.extractContext(entityAtts);
+        if (typeInformation.active) {
+            expressionAttributes = parser.processExpressionAttributes(typeInformation, typeInformation.active, context);
```
Comment (jmcanterafonseca) [inline lib/pluginsNgsi2/expressionPlugin.js]: line seems to be very long. check with a linter

File: lib/pluginsNgsi2/pluginUtils.js [hunk 12]
```diff
@@ -0,0 +1,171 @@
...
+ * For those usages not covered by the GNU Affero General Public License
+ * please contact with::daniel.moranjimenez@telefonica.com
+ *
+ * Modified work Copyright 2017 Atos Spain S.A
+ */
+
+'use strict';
+
+function createProcessAttribute(fn, attributeType) {
+    return function(attribute) {
```
Comment (jmcanterafonseca) [inline lib/pluginsNgsi2/pluginUtils.js]: new line not needed

File: lib/fiware-iotagent-lib.js [hunk 13]
```diff
@@ -116,6 +130,14 @@ function doActivate(newConfig, callback) {
         commandRegistry = require('./services/commands/commandRegistryMemory');
     }
 
+    if (checkNgsi2(newConfig)) {
+        logger.info(context, 'Using NGSIv2');
+        ngsi = ngsi2;
+        deviceService = deviceServiceNgsi2;
+        commands = commandsNgsi2;
+        subscriptions = subscriptionsNgsi2;
```
Comment (fgalan) [inline lib/fiware-iotagent-lib.js]: Note IOTAs don't uses any subscription... this seems to be over-implemented. The three NGSIv2 interactions that should be supported are only the ones described in #527 

1. Registration to CB on device provision time.
2. Send updates to CB (for active attributes).
3. Acts as Context Provider for queries/updates coming from CB (for passive attributes and commands).

Of those, only point 2 is achiavable now (as 1 and 3 are not implemented/specified yet). I'd suggest to reduce the scope just to that point. That should reduce the size of the PR and make it reviewable.

Not sure of the NGSIv1-based code is structured, but if the POST /v1/updateContext is modularized in just one point (where the verb + url + payload is composed) maybe is a matter of just adapting that part, creating a NGSIv2 version for that and switching to the new/old method depending on the parameter in config.js.

File: lib/services/common/domain.js [hunk 14]
```diff
@@ -54,7 +54,7 @@ function cleanDomain(domainToClean) {
 
 function generateDomainErrorHandler(domainToHandle) {
     return function domainErrorHandler(err) {
-        logger.error(context, err);
+        logger.error('aaaaaa'+context, err);
```
Comment (mrutid) [inline lib/services/common/domain.js]: Fix log message

File: lib/services/common/genericMiddleware.js [hunk 15]
```diff
@@ -108,6 +108,18 @@ function ensureType(req, res, next) {
 }
 
 /**
+ * Ensures the request type is one of the supported ones.
+ */
+function ensureTypeNgsi2(req, res, next) {
```
Comment (mrutid) [inline lib/services/common/genericMiddleware.js]: It's ensuring content type==JSON, not Ngsiv2

File: lib/services/northBound/restUtils.js [hunk 16]
```diff
@@ -148,8 +148,19 @@ function isTimestamped(payload) {
     return false;
 }
 
+function isTimestampedNgsi2(payload) {
```
Comment (mrutid) [inline lib/services/northBound/restUtils.js]: Please check "payload.hasOwnProperty(i)" when iterating over objects

File: lib/pluginsNgsi2/addEvent.js (hunk 6 above)
Comment (fgalan) [reply lib/pluginsNgsi2/addEvent.js]: In the case of an existing file, "Modified by:" would be better.

**Comment by dcalvoalonso** [comment]
Thanks for your prompt review and comments, I really appreciate the time and effort you have spent to give me feedback :).
I will try to clarify some points below:
* I will proceed as you propose and I will split the changes in several PRs to simplify the review process.
* I will add an ownership sentence as you propose instead of the Modified work Copyright statement:  Modified by:: Daniel Calvo - ATOS Research & Innovation 
* The approach I have followed is to try to maintain as clean as possible the original NGSIv1 code. For instance, if we talk about how to send updates for active attributes to the CB, I see that we have two options:
  1. To add a new parameter to sendUpdateValue function in ngsiService.js to choose between NGSIv1 and NGSIv2. This would imply modifying also other functions like generateNGSIOperationHandler. From my point of view, the resulting code would be more complex but of course, it is an option.
  2. To create a new version of ngsiService that implements specific code for NGSIv2. In fiware-iotagent-lib, as part of the activation and configuration loading, we check if NGSIv2 is used and use the corresponding ngsiService. In my eyes, this solution is less intrusive in the original code and the migration and support for new NGSIv2 related functionalities would be simpler. I have followed this alternative.
* Regarding plugins, I guess that the discussion is more or less similar to the previous point. Do we need a new plugin for NGSIv2? Well, we can add a new parameter in some functions to decide if we have to use NGSIv1 or NGSIv2 (for instance, updatePlugin function of timestampProcessPlugin.js) or just isolate the new specific code for NGSIv2 in a separate plugin. Both options are feasible and I am open to your proposals.
* Regarding subscriptions in IOTAs. The current version does implement subscriptions using NGSIv2:
  1. https://github.com/Fiware/iot.IoTagent-node-lib/blob/master/lib/services/ngsi/subscriptionService.js
  2. https://github.com/Fiware/iot.IoTagent-node-lib/blob/master/test/unit/ngsiService/subscriptions-test.js
  3. As you can see, I understand from the code that the idea is that a client using an IOTA can send a subscription request to the CB from the IOTA.
 * In general, what I have tried is to replicate all the functionalities that are implemented in the current version for NGSIv1 for NGSIv2.
* Regarding passing Travis-CI before merging... I completely agree with you... but if you check the job log you can see that tests failing are for NGSIv1. If you directly clone the master branch of the repository and run npm install and npm test, the output is the following:
```
263 passing (27s)
  4 pending
  20 failing
```

**Comment by fgalan** [comment]
With regards to approach and plugins, from my point of view having two plugins (one for NGSIv1 and another NGSIv2) or any other mechanism of "parallel" code could make senses in the case we want NGSIv1 and NGSIv2 at the end. However, note that NGSIv2 and NGSIv1 funcionality will coexits for short time (as we [dicussed in the root issue](https://github.com/telefonicaid/iotagent-node-lib/issues/527#issuecomment-337565075)). At the end, only NGSIv2 will remain. In this sense, I think that the complexity introduced in sendUpdateValue and other functions will be temporal, thus makes sense going on that line (as I understand new PR #556 does).

With regards to subscription, you are rigth and I was wrong :) I don't know why that code is there (maybe it comes from the old days, when it was not so clear how to integrate IOTA and CB) but it is not used. I'd create and issue about removing that code. There is not need of migrating that operation to NGSIv2.

With regards to travis, currently master is ok at https://travis-ci.org/telefonicaid/iotagent-node-lib/branches. Maybe the problems is the one that @AlvaroVega point out at https://github.com/telefonicaid/iotagent-node-lib/pull/556/#discussion_r147087022

**Comment by dcalvoalonso** [comment]
@fgalan, I will proceed then as you suggest for plugins, I think that it's a good approach! :)

Regarding travis, I have seen that @AlvaroVega has the same problem in #557  . As you can see, the unit tests failing are exactly the same. :(

**Comment by fgalan** [comment]
> With regards to subscription, you are rigth and I was wrong :) I don't know why that code is there (maybe it comes from the old days, when it was not so clear how to integrate IOTA and CB) but it is not used. I'd create and issue about removing that code. There is not need of migrating that operation to NGSIv2.

Issue created https://github.com/telefonicaid/iotagent-node-lib/issues/558

**Comment by fgalan** [comment]
Finally, the work was done in PR #556 . I understand this PR can be closed. Is that correct, @dcalvoalonso ?

**Comment by dcalvoalonso** [comment]
Yes, I agree with you! :+1:
